File: src/322.零钱兑换.rs
/*
 * @lc app=leetcode.cn id=322 lang=rust
 *
 * [322] 零钱兑换
 */
struct Solution;
// @lc code=start
impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let amount = amount as usize;
        const FAIL: i32 = -1;

        let mut dp = vec![0; amount + 1];
        for target in 1..=amount {
            dp[target] = coins
                .iter()
                .filter(|&&change| change as usize <= target)
                .map(|&change| match dp[target - change as usize] {
                    FAIL => FAIL,
                    count => count + 1,
                })
                .filter(|&count| count > 0)
                .min()
                .unwrap_or(FAIL);
        }
        dp[amount]
    }
}
// @lc code=end

#[test]
fn test_solution() {
    macro_rules! test {
        ($v:expr, $amount:expr, $ans:expr) => {
            assert_eq!(Solution::coin_change($v, $amount), $ans);
        };
    };
    test!(vec![1, 2, 5], 11, 3);
    test!(vec![2], 3, -1);
}

File: src/322.零钱兑换.py
#
# @lc app=leetcode.cn id=322 lang=python3
#
# [322] 零钱兑换
#
from prelude import *
# @lc code=start


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [-1 for _ in range(amount+1)]
        dp[0] = 0
        for i in range(amount):
            if dp[i] == -1:
                continue
            # i => ...
            for coin in coins:
                if i + coin <= amount:
                    if dp[i+coin] == -1:
                        dp[i+coin] = dp[i] + 1
                    else:
                        dp[i+coin] = min(dp[i]+1, dp[i+coin])

        return dp[amount]
# @lc code=end
